### DISCUSSION TITLE
Imbalance between the two sides

### DISCUSSION THREAD

**Discussion by peyre** [raceintospace/raceintospace#771] [Ideas]
I've commented in my guide that overall, the Soviets are at a bit of a disadvantage in RIS.  However I'm starting to think I may have understated that point.  Lately I've noticed that when I play as the Soviets against the AI I usually lose - but if I play as the Americans I win more often than the AI does.  I wonder if there's much of anything we could do to try to balance that...ideally, things that would accentuate the differences between them and/or improve the game's realism.

One idea that comes to mind is lowering the cost of launch pads by 5MB for the USSR (maybe also the cost to repair them).  Another is more ambitious: allowing the Soviets to conduct unmanned LM Tests; IRL the Soviets actually tested their LM robotically.  You can't just add missions though; we'd probably have to poach another mission type to implement this.

**Comment by hweimer** [comment]
Interesting observation. I ran 100 games with the AI playing against itself, 63 times the U.S. won and 37 times the Soviets won (there were no ties).  The standard error is about 5 wins, so it indeed looks like there is some kind of imbalance. But then I switched to the basic model, which is supposed to be identical for both sides. Then, I still had the U.S. winning 61 times and the Soviets winning 39 times. So, I suppose that the imbalance either comes from some asymmetry in the AI code or from a bug favoring the U.S. side. I was thinking that it might have to do something with the Soyuz lunar landing because it provides an asymmetry even in the basic model, but the AI appears to never fly them.

I will run more simulations to reduce the statistical error and see what I get.

**Reply by NooneAtAll3-2** [reply]
do you by chance still have the code for self-playing the Ai? I'd love to tinker with it

**Comment by peyre** [comment]
Wow!!  Good work--I wouldn't have thought to run that sort of test, I don't think.  (Frankly I was just hoping not to get a reaction like "Well, everyone likes playing the Americans better anyway, so why would we change it?")

**Comment by peyre** [comment]
Hey I tried that this morning, by modifying line 400 in prefs.cpp.  But you have to keep pressing keys through every single prestige first anyone gets!  It'd be nice if the Esc key would close out the Prestige First video.  Trying to run these tests is especially annoying since with both players as AI it plays each mission two or three times!  (And some of them show stuff that shouldn't be there, like a spacewalk on a probe landing.)

Hey, I wonder if we should change the Preferences screen to allow users to do that, set both sides to AI and watch 'em go.  Some games let you do that.

**Reply by hweimer** [reply]
I have made a few modifications (removing replays, for instance) to speed things up. With these, an AI-vs-AI game takes about 15 seconds. I can put my changes into a separate branch in case you are interested.

**Comment by peyre** [comment]
Sure enough!  I just ran 10 tests (it's slow and dull having to hold down a key at every prestige first movie!), in the Historical Model, and yep, the US won 7 times, the USSR only 3.  I'm glad you thought to try running AI vs. AI to eliminate the variable of the human player.

**Comment by rnyoakum** [comment]
There are a number of different factors that are country-dependent, regardless of basic/historical mode. My guess, off the top of my head, is that it's probably due to budget. The USA and USSR use different random budget modifiers which advantage the USA. That's the sort of thing which can snowball. Check REvent() in news_suq.cpp.

**Reply by hweimer** [reply]
That's funny because from my reading of the comments surrounding the reading of the modifier table and the layout of the file (there are three different tables), it looks that the modifiers were originally meant to be not for the different sides but for different player *levels*. So, I consider this to be a bug.

**Comment by peyre** [comment]
That's a good point Ryan.  I've noticed that catastrophic failures are more time-consuming and expensive for the Soviets to recover from, because their Max R&D figures are lower; maybe that's the key to this, or part of it.  

What to do about it I don't know.  One idea that comes to mind is that Max R&D following a catastrophic failure is the higher of the two sides' initial Max R&D figures (meaning effectively the US's figure).  So for instance Soyuz has an 85% Max R&D at the start of the game, and Apollo 89%.  A cat failure of Soyuz would give it a Max R&D of 89% rather than 85%.  That would even out a big part of the imbalance of the game, though it feels contrived.

If we _were_ to implement something like that, we'd probably want to set a variable ("CatMaxRD"?) at the start of the game equal to the higher Max R&D figure, which both sides would drop to in the event of a catastrophic failure: 83% for Gemini/Voskhod, 89% for Apollo/Soyuz, etc.  This would also have the advantage of fixing #545.  I'm not sure this would work well for the minishuttles though, since XMS-2 has a much higher Max R&D than Lapot.

Hendrik, that sounds really helpful for testing purposes, thought I don't think I need it myself at the moment.  I wonder if it might be handy to have a Testing branch for testing AI vs. AI.

**Reply by peyre** [reply]
Another possibility, which might be less obtrusive to users, might be if the Soviets took half as much of a hit from catastrophic failures, presumably because they're less subject to the approval of the public.

**Comment by hweimer** [comment]
I ran now 1,000 games for both the basic and the historical model.
Basic model: 539 U.S. wins, 461 Soviet wins.
Historical model: 638 U.S. wins, 362 Soviet wins.

So, it looks like that the similarity between the basic and the historical model was a statistical fluke. There is likely some imbalance in the basic model probably due to the budget modifiers as Ryan pointed out, but it is nowhere as severe as in the historical model.

But to some extent, it brings up the usual question of gameplay vs historical accuracy. The space race was imbalanced because the U.S. had a larger economy (partly but not fully explained by different population sizes). But more than 60% sounds too much. Chess, for instance, has white winning about 55% of the time according to recent statistics.

**Comment by peyre** [comment]
That's enlightening!  A sample size of a thousand is much better.  So, it looks like the Basic Model is much closer to 50/50 than we'd thought.  This new info suggests a couple things.

1. It seems the game still runs about 54/46, so something must be putting a little bit of a finger on the scale.  It still might be random variation at work here, but I'm inclined to think there's a little something at work.  Is it maybe that even though costs are the same, the Soviets still tend to have a lower budget than the Americans?
2. We still have a significant problem in the Historical Model.  I know it sort of models the historical fact that the Soviets were at a disadvantage late in the Space Race, but this may be one of those cases where historical realism hinders gameplay.

**Reply by hweimer** [reply]
Concerning 1, I think it makes sense that the Soviets have a harder time if they are using level 2 budget modifiers compared to the level 1 modifiers for the U.S. side. Concerning 2, I want to run a few more simulations (e.g., with different levels) to get a better idea how things could be improved.

**Reply by peyre** [reply]
I'm not sure I know what you mean about the budget modifiers.

**Reply by hweimer** [reply]
As I wrote to Ryan in https://github.com/raceintospace/raceintospace/discussions/771#discussioncomment-5768032, it seems that originally, the budget modifiers were for different player levels and not for the different sides. So, the U.S. plays with what used to be the level 1 budget modifiers, while the Soviets get the level 2 modifiers, which is likely to be a disadvantage. Not sure why this got changed at some point.

I also found another source of imbalance even in the basic model. If two missions are flown in the same month, the player with the larger budget goes first. However, if the budgets are the same (not unlikely after a few turns), the U.S. mission is much more to be flown first.

**Reply by peyre** [reply]
Good spotting!  I assume these are items we'll want to address at some point?

I wonder what the rationale was for having the player with the higher budget go first.  That seems unbalancing.

I've mentioned this on the project's Mastodon account.  I think Michael is one (of two!) subscribers; maybe he'll chime in with some background info.

**Reply by hweimer** [reply]
Yeah, we should address these points in some way. Right now, I'm busy with fixing all the bugs I encounter during AI-vs-AI games. I can't even get to 1,000 games in a row because the game always finds a new way to crash. Luckily, everything is reproducible, so this is really a great way to improve things.

**Reply by rnyoakum** [reply]
I think the original idea was that for a historic simulation, the game would reflect the greater economic engine of the United States, as a counterbalance to things like the Soviet access to a Soyuz lunar landing. That's a difference between the sides even in Basic, and I think we should compile a listing of all the differences. However, clearly the two sides are not balanced and any advantage the Soviets get from the Soyuz landing are outweighed by other factors.

I see the following options for budgeting, and can implement any of them:

1. Leave the budgets as they are
2. Switch both USA and USSR to using the Level 1 budgets modifiers
3. Switch the USA and USSR to using the same Level-appropriate modifiers
4. Implement  (2) or (3) for the Basic model, but leave them as-is for the historic model to represent the USA's stronger economic engine
5. Create different budget modifier tables for the USA and USSR so there remains a difference between the USA and USSR, but not as big

Personally, I recommend changing the USA and USSR to use the same budget modifiers (option 2 or 3) for the Basic model, because that's designed to be a level playing field for PvP. As to historic mode, while sentiment says the game should incorporate the greater budget power of the USA logic argues money is probably _the_ most influential element, and the USA doesn't need the additional advantage. So between option 2 and 3, I would probably go with option 2; Level-based budget modifiers appear to have been dropped because they made the game too hard.

Leon?


P.S. Another option is to use the Level 1 budget modifiers (basic & historic) but use Level-dependent modifiers when playing in 'Classic' mode.

**Reply by peyre** [reply]
Hendrik, I'm glad to see you finding and fixing all these obscure bugs.  This is sort of "unsung hero" stuff because these are errors that won't happen very often, so you're eliminating the sort of thing that would often go unfixed since they happen infrequently enough that they'd often go unreported.

Good rundown Ryan.  Completely in agreement with evening the playing field for the Basic Model.  That's what it's for, after all.  The Historical Model is more complex.  You're right about the importance of money.  Leveling the budget modifiers is clearly an option.  Another would be to adjust hardware costs etc. to compensate for the USSR's lower budget--not a very good option, probably: most people probably wouldn't welcome lots of changes to those.  

Your option 5, as I understand it, would be to try to model the differing impacts of failures etc. caused by the Soviet program being secret.  Failures would be less damaging, and maybe successes would be a bit less helpful, maybe.  This is an intriguing suggestion, and I've toyed with the idea myself, but it would be complicated to implement and would need a lot of testing to protect game balance.  For now at least, I'm inclined to just go with 2, putting both sides on the same budget modifiers.  Maybe we could consider 5 for future releases.

**Reply by hweimer** [reply]
Thank you for your kind words. Right now, my bug squashing spree is more driven by my hope to complete 1,000 games without crashing than by helping out other players. But I have to say that this seems to be a lot more efficient in finding bugs compared to, say, standard code coverage procedures.

**Comment by hweimer** [comment]
I'm now down to about 0.1s per game by essentially disabling all graphics output and by keeping all deserialized data in memory. I'm also seeing a lot of instances of #607, but not when running the game in a debugger. *sigh*

**Comment by hweimer** [comment]
Now this is surprising: if I restore the budget modifiers to the previous behavior (i.e., depending on player level), it's actually the Soviet side that has a significant advantage in the basic model (55.1% wins). The average final budget is also a bit higher for the Soviets, but it's difficult to say whether this a cause or an effect.

What other sources of imbalance do we have? 'Nauts have different skill points, but I somehow doubt that this effect is large enough. Soyuz L.L. cannot be the cause either, because the AI doesn't fly it if I'm not mistaken.

**Reply by peyre** [reply]
Odd.  Spacepeople, at least, shouldn't be causing this imbalance.  The US has better CA skills, the Soviets a bit better on all other skills--but that's still a net win for the US.  I can't think what might be advantaging the USSR in the Basic Model.

**Reply by rnyoakum** [reply]
I'm wondering how random the mission ordering is. It looks as if the cmp_order function that orders the missions got accidentally modified. In the case of a perfect tie between two missions (same date, budget) the random choice between them uses brandom(9) instead of brandom(2). My guess is this gives the USSR side a 8-1 advantage on getting the first shot at the Orbital Satellite, which would produce a budgeting disparity.

**Reply by peyre** [reply]
That would certainly be worth fixing.  If all things are equal, including on Orbital Satellite, it should be 50/50.

**Reply by rnyoakum** [reply]
Yeah. The value was accidentally changed when some text colors were being edited. I'm not certain how much of an effect it's having or if it even is having one - it's less effort to fix it than puzzle out the consequences. But you got my guess.

**Reply by peyre** [reply]
So, this was a change we introduced accidentally?  (If so, I won't add this to the changelog, since it's something we introduced.)

**Reply by rnyoakum** [reply]
I wouldn't mention anything in the changelog, because it's fixing a problem introduced last year. The error occurred in commit b2c1ae568a56d5f3922c60bc4f39483a32a99896. The commit made a bunch of color changes to text, and this one value got modified by accident, for no apparent reason.  I corrected it last night.

**Reply by peyre** [reply]
Ah, good thanks.  Yes, the changelog says we brought the game up to where it was at 1.1 on SourceForge, and below are the changes we've made to it beyond that.  So good, I won't mention this.

**Reply by hweimer** [reply]
I had checked this already, the bug in the random mission ordering was not affecting the chances. Strangely, I cannot reproduce the increased odds for the Soviets any longer. Must have been something else that I did. Anyway, the two sides are still not balanced in the basic model. However, the chances both for the first orbital satellite and the first manned mission are pretty much evenly split, so the deviation must occur later in the game.

**Reply by peyre** [reply]
I've always suspected the Soviet player is a bit more likely to get the newscast that gives you (and requires) female cosmonauts.  I wonder if that extra bit of prestige might help the USSR?  (_Or_ that might be confirmation bias on my part.  Maybe it's no more likely for either side.)

**Reply by hweimer** [reply]
I don't think it's the female 'nauts, but your suspicion concerning the newscasts points into an interesting direction. Contrary to what I thought, newscasts are exclusive to one player, so if you get a certain event card, the other player cannot receive it as well during the entire game. This indeed provides an advantage to the U.S. in case of a game-changing event card (the half-off sale comes to my mind) because it takes the first draw. However, it does not seem to play a major role in AI-vs-AI games because money is less of a concern due to the AI money cheat. Forcing both players to see the same events still has the U.S. ahead at 53.7% after 10,000 games.

But I think that the notion of getting bad events meaning your opponent getting better ones is worth discussing, especially in the light of the discussions we already had in #542.

**Reply by peyre** [reply]
It just occurred to me, on the other hand, that if this also would mean that if the US picked a bad card, the Soviet player wouldn't get it, correct?  In that case, between good and bad cards this may be a wash.

Perhaps newscasts shouldn't be exclusive like that?  Like maybe the game shouldn't let both sides get the same card in the same turn, but if one side gets a half-off sale, the other side _should_ be able to pick it up in a later draw.

That leads me to wonder if the game really should predetermine the newscasts at the start of the game.  Maybe it would be better if it did them randomly over time?  That would afford us some flexibility in regards to #542.
